### PR TITLE
Dealing with long time length readings

### LINKED ISSUE #665: Meter-by-meter logic for serving compressed data
### Is your feature request related to a problem? Please describe.

Fictitious data is displayed when the underlying reading data has a lower reading frequency than the frequency that the compressed readings algorithm tries to serve. For example, if the compression algorithm tries to serve data from the now materialized hourly view when the underlying data is 4-hourly data. This compression serves fictitious data because there is not actually any hourly data; instead a constant or linear approximation is applied. As the algorithm is now the only "true" compressed data would be from the daily view.

### Describe the solution you'd like

One idea to resolve this would be to track the reading frequency onto the meter as a new column. This could either be provided from the administrators when they first create the meters or be estimated from the first batch of readings from the data. Then the compression algorithm would determine the appropriate level of compression for serving data for certain time intervals.

### Describe alternatives you've considered

OED does not indicate whether the data displayed is compressed or not. We could instead visually indicate the level of compression to the user when they read the graph. This would be less misleading since they should also know the actual reading rate of the meter and understand that the compression is not displaying real data. 

### Additional context
With the addition of PR #660 this only becomes an issue if the meter reads slower than 1 reading per hour. For electricity, this is less of an issue since most meters read data every 15-minutes. As OED generalizes this might become an issue for certain energy sources.

### LINKED ISSUE #806: dealing with low frequency readings
### Is your feature request related to a problem? Please describe.

If a meter only acquires data at longer time frames then it is possible OED will graph it at a higher frequency by adding points. An example would be a water meter from the utility bill that only had a data point each month. OED would graph it daily and add 28-31 points for each day between the actual month points. This can happen, in general, if the reading frequency is longer than 1 hour since we have hourly and daily views.

### Describe the solution you'd like

The DevDocs longReadings.md discusses this in detail and proposes a solution. Basically, we use raw/meter reading when possible and then go to hourly and then daily.

### Describe alternatives you've considered

Unsure but other ideas are possible

### Additional context

This will address issue #519 that was partly addressed when the minutely readings were removed so it is being closed.

### PR DESCRIPTION
**Pull request by kirtmd**
# Description

Dealing with long time length readings

Fixes #665, Fixes #806.

Changed the way that meter data is returned, previously OED would use daily and hourly readings before going to raw readings which lead to showing readings that didn't exist. With this code, OED starts by checking raw readings, then hourly and daily last to make sure that these issues don't persist. We added a reading frequency column to the meters table to help us do this. The function loops through each meter that is input because we need to know the frequency of each meter specifically to return it's proper reading interval.

(Check the ones that apply by placing an "x" instead of the space in the [ ] so it becomes [x])

- [ ] Note merging this changes the node modules
- [x] Note merging this changes the database configuration.
- [ ] This change requires a documentation update

## Checklist

(Note what you have done by placing an "x" instead of the space in the [ ] so it becomes [x]. It is hoped you do all of them.)

- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [ ] I have removed text in ( ) from the issue request

## Limitations

Issue #665 is still not completely finished, this covers the meter section of the issue but the group part is still TODO, you can read more about it here: https://github.com/OpenEnergyDashboard/DevDocs/blob/main/longReadings.md

### CODE REVIEW

File: src/server/migrations/0.8.0-1.0.0/sql/meter/add_meter_reading_frequency.sql
Comment (huss) [inline src/server/migrations/0.8.0-1.0.0/sql/meter/add_meter_reading_frequency.sql]: I think this comment was copied from another file and needs to be updated.

File: src/server/migrations/0.8.0-1.0.0/sql/meter/add_meter_reading_frequency.sql
Comment (huss) [inline src/server/migrations/0.8.0-1.0.0/sql/meter/add_meter_reading_frequency.sql]: There is a default value in the meters creation SQL that should also be here.

File: src/server/sql/reading/create_reading_views.sql
Comment (huss) [inline src/server/sql/reading/create_reading_views.sql]: Since add is the current meter id, should the name reflect that. meter_id is okay but seems close to meter_ids. Open to ideas but what about current_meter_id?

File: src/server/sql/reading/create_reading_views.sql
Comment (huss) [inline src/server/sql/reading/create_reading_views.sql]: While you nicely put comments below, I think some more, esp. on what the next few lines are doing would help future developers.

File: src/server/sql/reading/create_reading_views.sql
Comment (huss) [inline src/server/sql/reading/create_reading_views.sql]: Should this commented out line be removed?

File: src/server/sql/reading/create_reading_views.sql
Comment (huss) [inline src/server/sql/reading/create_reading_views.sql]: I cannot convince myself that the logic precisely does what is wanted but is close. It seems to assume that the requesteed_interval is an even number of days and I don't think that is the case. I think this since I thought EXTRTACT DAY only returns whole days. I also think we could have an issue with frequency if someone entered it in the general Postgres syntax of '4 days' rather than '96:00:00' since it does not deal with days. A quick test found this did not work with the current code.

To do this in a general way is probably desirable given that there are variety of ways to enter these values into Postgres. I'm hoping there is a way to do this but I have not investigated.

**Review by huss — CHANGES_REQUESTED** [review]
Thanks to @kirtmd for putting in their first PR. I have left some comments to think about and address.

I did modest testing and found this generally works as desired. More substantial testing will be needed after it is finalized and before merging.

I also noticed it has conflicts with development over other changes that need to be resolved. I can help with that if you want.

Also see a comment on this PR (about to do) about other changes needed in OED before this can be deployed.

**Comment by huss** [comment]
Working with this PR and thinking some more lead me to realize that the following changes are needed in OED before this can be deployed:

- Changing the way meters gets data impact groups (that use the new code for each meter in a group) so the group code needs to be updated. It needs to be the case that all meters are getting readings at the same frequency so they can be summed. We have to decide which of the possible meter frequencies to use but the one with the fewest points would be the conservative choice. It also needs to exclude raw. All of this will likely mean changes/new parameters for the line meter function.
- The meter pages need to be changed to allow interaction with the meter frequency of reading.
- The CSV meter import needs to work with the meter frequency of reading (might work okay but need to check).
- Many automated test cases that looks at line readings returned will need to be updated to reflect the new values returned.
- The test data and website data need to be updated to include the correct frequency of reading for each meter. This will also require modifying the meter input code to handle this.

Given that OED v1.0 should happen soon and the changes should be done before this PR is included in an OED release, this PR is on hold for merging to make sure either it happens after v1.0 or all changes are in place before merging.

Please note this does not reflect on the work done for the PR but issues outside of this work that must be accomplished. This PR can still progress with changes and be ready to merge once all the other work is done.

**Comment by huss** [comment]
I have worked on the necessary changes. The important ones to note are:

- The comments on code in the PR are all addressed.
- Each meter correctly calculates the number of points for its readings in DB call. In the past it was for all meters.
- The DB now uses passed parameters to determine when to switch between raw, hourly and daily. It uses 1440 points for now.
- Groups now work with the new code where all meters return the same frequency so they can be added and raw is excluded.
- The preferences page allows setting of a default meter frequency reading. The old environment variable is removed.
- The meter page allows setting of the reading frequency.
- Changes integrated in all graphing pages.
- All meter creation and tests were updated for the changes in this PR.
- Changed so meter create/update returns the id and reading frequency. This is routed back to the client to set state. This means a reload of all meter state is no longer needed.
- A bug was fixed that messed up the meter unit choice on the meters page.
- A bug was fixed where line state was sometimes unset and caused an error.
- The test data for developers and web were updated for the new code. This includes changing the meter loading of test data to be by key:value pairs. The data now sets the reading frequency & area but not area for web since need to figure out what to do. Note groups have not been converted so their area is not set.

Note that a default of 15 minutes is used for reading frequency. A lower value was used since it will return fewer points and be faster if the admin does not set.

With these changes I think this PR is ready to consider for merging into OED 1.0.

**Review by spearec — APPROVED** [review]
I've looked over the code and everything looks good. I don't have a perfect understanding of the SQL logic used, but the comments were enough to give me a passable understanding of what is going on.

I wasn't able to test the changes much, but simply graphing the test4DaySin meter in OED is enough to immediately see that the changes seem to fix the issue at least in this case.

The only thing I noted that might be an issue is that reading_line_accuracy is not defined in a migration, but I'm not sure if this is needed depending on when the enum is created in OED.

**Comment by huss** [comment]
> The only thing I noted that might be an issue is that reading_line_accuracy is not defined in a migration, but I'm not sure if this is needed depending on when the enum is created in OED.

Thanks for catching that. A coming commit fixes it.

**Comment by huss** [comment]
Thanks to @spearec for jumping in and doing this review. I am taking it as approval. I should have addressed the one comment. It should be ready to merge.

**Pull request merged**
